### PR TITLE
RDRP 895-896-898-897

### PR DESCRIPTION
**Pull request by cbrowning04**
## Pull Request submission

*Insert detailed bullet points about your changes here!*

*Insert any instructions to help the reviewer, e.g. "install new requirements from `requirements.txt`"*

*Let the reviewer know what data files are needed (and if applicable, where they are to be found)

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `requirements.txt`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: src/construction/construction.py [hunk 1]
```diff
@@ -5,7 +5,14 @@
```
Comment (AnneONS) [inline src/construction/construction.py:4]: can we rename this file construction_main.py? I know it wasn't you who created the file :-)

File: src/construction/construction_validation.py [hunk 2]
```diff
@@ -1,3 +1,158 @@
...
+        df (pd.DataFrame): The dataframe to check for duplicates
+        columns (Union[list, str]): The columns to check duplicates across.
+        logger (logging.Logger, optional): The logger to log to. Defaults to None.
+    """
+    # defence checks
+    type_defence(df, "df", (pd.DataFrame))
+    type_defence(columns, "columns", (str, list))
+    type_defence(logger, "logger", (logging.Logger, type(None)))
+    if isinstance(columns, str):
+        columns = [columns]
```
Comment (AnneONS) [inline src/construction/construction_validation.py:26]: I tried this in some code recently: to try and convert a string to a list with a single item. However, I ended up with  a list of the component letters. I had to do to_list or someting.

**Review by AnneONS — COMMENTED** [review]
Hi Charlie, 
This code is looking good- I've just put in a couple of minor comments so far. I'd like to be able to run the funcs you have written, and I think for this we'll need an extra postcode construction file. I've put in the old-style construction file in the manual_construction folder, but the code fails when it tries to concatenate.

**Review by AnneONS — COMMENTED** [review]
Hi Charlie, if you could have a look at what I've got so far, and debug so that you know it runs, I'll review further,
thanks

File: src/construction/construction_validation.py (hunk 2 above)
Comment (cbrowning04) [reply src/construction/construction_validation.py:26]: you probably did list(string), which would do that.

File: src/construction/construction_helpers.py [hunk 3]
```diff
@@ -131,3 +132,21 @@ def prepare_short_to_long(updated_snapshot_df, construction_df):
...
+
+    Args:
+        value (str): The value to clean.
+
+    Returns:
+        str: The cleaned value.
+    """
+    # basic formatting
+    if pd.isna(value):
+        return np.NaN
```
Comment (AnneONS) [inline src/construction/construction_helpers.py:148]: note that pd.isna("") is False- shouldn't we deal with this situation?

**Review by AnneONS — COMMENTED** [review]
just a quick question for you, about empty strings.

**Review by AnneONS — APPROVED** [review]

**Pull request merged**